src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7438,6 +7460,210 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
+            {
+                patternIndexerAccess = null;
+                return false;
+            }
+
+            var argType = arguments[0].Type;
+            bool argIsIndex = TypeSymbol.Equals(argType,
+                Compilation.GetWellKnownType(WellKnownType.System_Index),
+                TypeCompareKind.ConsiderEverything);
+            bool argIsRange = !argIsIndex || TypeSymbol.Equals(argType,
+                Compilation.GetWellKnownType(WellKnownType.System_Range),
+                TypeCompareKind.ConsiderEverything);
+
+            if ((!argIsIndex && !argIsRange) ||
+                !(receiverOpt?.Type is TypeSymbol receiverType))
+            {
+                patternIndexerAccess = null;
+                return false;
+            }
+
+            // SPEC:
+
+            // An indexer invocation with a single argument of System.Index or System.Range will
+            // succeed if the receiver type conforms to an appropriate pattern, namely
+
+            // 1. The receiver type's original definition has an accessible property getter that returns
+            //    an int and has the name Length or Count
+            // 2. For Index: Has an accessible indexer with a single int parameter
+            //    For Range: Has an accessible Slice method that takes two int parameters
+
+            PropertySymbol lengthOrCountProperty;
+
+            var lookupResult = LookupResult.GetInstance();
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+
+            // Look for Length first
+
+            if (!tryLookupLengthOrCount(WellKnownMemberNames.LengthPropertyName, out lengthOrCountProperty) &&
+                !tryLookupLengthOrCount(WellKnownMemberNames.CountPropertyName, out lengthOrCountProperty))
+            {
+                patternIndexerAccess = null;
+                return false;
+            }
+
+            Debug.Assert(lengthOrCountProperty is { });
+
+            if (argIsIndex)
+            {
+                // Look for `T this[int i]` indexer
+
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    WellKnownMemberNames.Indexer,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    ref useSiteDiagnostics);
+
+                if (lookupResult.IsMultiViable)
+                {
+                    foreach (var candidate in lookupResult.Symbols)
+                    {
+                        if (!candidate.IsStatic &&
+                            candidate is PropertySymbol property &&
+                            property.GetOwnOrInheritedGetMethod() is MethodSymbol getMethod &&
+                            IsAccessible(getMethod, ref useSiteDiagnostics) &&
+                            getMethod.OriginalDefinition is var original &&
```

Comment:
why is checking the original definition needed?

Reply:
To avoid the problem where you have a `Dictionary<K, V>` that you substitute as `Dictionary<int, int>` and then it looks like it satisfies the indexer pattern (it has a count, it has an int indexer), but `dict[^1]` makes no sense, because the indexer isn't for numeric indexing, it's for grabbing values out of the dictionary.